src/task/future.rs

```diff
@@ -0,0 +1,393 @@
+{
+    drop(task_cell(this as *const Task<Remote, Extras>))
+}
+
+unsafe fn wake_impl<Remote, Extras>(task_cell: &TaskCell<Remote, Extras>)
+where
+    Remote: RemoteSpawn<TaskCell = TaskCell<Remote, Extras>>,
+{
+    let task = &task_cell.0;
+    let mut status = task.status.load(SeqCst);
+    loop {
+        match status {
+            IDLE => {
+                match task
+                    .status
+                    .compare_exchange_weak(IDLE, NOTIFIED, SeqCst, SeqCst)
+                {
+                    Ok(_) => {
+                        task.remote.spawn(clone_task(&**task));
+                        break;
+                    }
+                    Err(cur) => status = cur,
+                }
+            }
+            POLLING => {
+                match task
+                    .status
+                    .compare_exchange_weak(POLLING, NOTIFIED, SeqCst, SeqCst)
+                {
+                    Ok(_) => break,
+                    Err(cur) => status = cur,
+                }
+            }
+            _ => break,
+        }
+    }
+}
+
+#[inline]
+unsafe fn wake_raw<Remote, Extras>(this: *const ())
+where
+    Remote: RemoteSpawn<TaskCell = TaskCell<Remote, Extras>>,
+{
+    let task_cell = task_cell(this as *const Task<Remote, Extras>);
+    wake_impl(&task_cell);
+}
+
+#[inline]
+unsafe fn wake_ref_raw<Remote, Extras>(this: *const ())
+where
+    Remote: RemoteSpawn<TaskCell = TaskCell<Remote, Extras>>,
+{
+    let task_cell = ManuallyDrop::new(task_cell(this as *const Task<Remote, Extras>));
+    wake_impl(&task_cell);
+}
+
+#[inline]
+unsafe fn task_cell<Remote, Extras>(task: *const Task<Remote, Extras>) -> TaskCell<Remote, Extras>
+where
+    Remote: RemoteSpawn<TaskCell = TaskCell<Remote, Extras>>,
+{
+    TaskCell(Arc::from_raw(task))
+}
+
+#[inline]
+unsafe fn clone_task<Remote, Extras>(task: *const Task<Remote, Extras>) -> TaskCell<Remote, Extras>
+where
+    Remote: RemoteSpawn<TaskCell = TaskCell<Remote, Extras>>,
+{
+    let task_cell = task_cell(task);
+    mem::forget(task_cell.0.clone());
+    task_cell
+}
+
+/// [`Future`] task runner.
+pub struct Runner<Spawn> {
+    _phantom: PhantomData<Spawn>,
+}
+
+impl<Spawn> Default for Runner<Spawn> {
+    fn default() -> Runner<Spawn> {
+        Runner {
+            _phantom: PhantomData,
+        }
+    }
+}
+
+impl<L, R, Extras> crate::Runner for Runner<L>
+where
+    L: LocalSpawn<TaskCell = TaskCell<R, Extras>, Remote = R>,
+    R: RemoteSpawn<TaskCell = TaskCell<R, Extras>>,
+{
+    type Spawn = L;
+
+    fn handle(&mut self, _local: &mut L, task_cell: TaskCell<R, Extras>) -> bool {
+        let task = task_cell.0;
+        unsafe {
+            let waker = ManuallyDrop::new(waker(&*task));
+            let mut cx = Context::from_waker(&waker);
+            loop {
+                task.status.store(POLLING, SeqCst);
+                if let Poll::Ready(_) = Pin::new(&mut *task.future.get()).poll(&mut cx) {
```

Comment:
Why not `(&mut *task.future.get()).as_mut().poll(&mut cx)`?

Reply:
I think there's no difference. I'll change if you prefer it.

Reply:
It's confusing to create a new `Pin` when the inner type is `Pin` already.